extras/rya.indexing.pcj/src/main/java/org/apache/rya/indexing/pcj/storage/mongo/MongoPcjDocuments.java

```diff
@@ -0,0 +1,438 @@
+import org.openrdf.model.impl.URIImpl;
+import org.openrdf.query.BindingSet;
+import org.openrdf.query.MalformedQueryException;
+import org.openrdf.query.QueryEvaluationException;
+import org.openrdf.query.QueryLanguage;
+import org.openrdf.query.TupleQuery;
+import org.openrdf.query.TupleQueryResult;
+import org.openrdf.query.impl.MapBindingSet;
+import org.openrdf.repository.RepositoryConnection;
+import org.openrdf.repository.RepositoryException;
+
+import com.mongodb.MongoClient;
+import com.mongodb.client.FindIterable;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.util.JSON;
+
+/**
+ * Creates and modifies PCJs in MongoDB. PCJ's are stored as follows:
+ *
+ * <pre>
+ * <code>
+ * ----- PCJ Metadata Doc -----
+ * {
+ *   _id: [table_name]_METADATA,
+ *   sparql: [sparql query to match results],
+ *   cardinality: [number of results]
+ * }
+ *
+ * ----- PCJ Results Doc -----
+ * {
+ *   pcjName: [table_name],
+ *   auths: [auths]
+ *   [binding_var1]: {
+ *     uri: [type_uri],
+ *     value: value
+ *   }
+ *   .
+ *   .
+ *   .
+ *   [binding_varn]: {
+ *     uri: [type_uri],
+ *     value: value
+ *   }
+ * }
+ * </code>
+ * </pre>
+ */
+public class MongoPcjDocuments {
+    public static final String PCJ_COLLECTION_NAME = "pcjs";
+
+    // metadata fields
+    public static final String CARDINALITY_FIELD = "cardinality";
+    public static final String SPARQL_FIELD = "sparql";
+    public static final String PCJ_ID = "_id";
+    public static final String VAR_ORDER_ID = "varOrders";
+
+    // pcj results fields
+    private static final String BINDING_VALUE = "value";
+    private static final String BINDING_TYPE = "uri";
+    private static final String AUTHS_FIELD = "auths";
+    private static final String PCJ_NAME = "pcjName";
+
+    private final MongoCollection<Document> pcjCollection;
+    private static final PcjVarOrderFactory pcjVarOrderFactory = new ShiftVarOrderFactory();
+
+    /**
+     * Creates a new {@link MongoPcjDocuments}.
+     * @param client - The {@link MongoClient} to use to connect to mongo.
+     * @param ryaInstanceName - The rya instance to connect to.
+     */
+    public MongoPcjDocuments(final MongoClient client, final String ryaInstanceName) {
+        requireNonNull(client);
+        requireNonNull(ryaInstanceName);
+        pcjCollection = client.getDatabase(ryaInstanceName).getCollection(PCJ_COLLECTION_NAME);
+    }
+
+    private String getMetadataID(final String pcjName) {
+        return pcjName + "_METADATA";
+    }
+
+    /**
+     * Creates a {@link Document} containing the metadata defining the PCj.
+     * @param pcjName - The name of the PCJ. (not null)
+     * @param sparql - The sparql query the PCJ will use.
+     * @return The document built around the provided metadata.
+     * @throws PCJStorageException - Thrown when the sparql query is malformed.
+     */
+    public Document getMetadataDocument(final String pcjName, final String sparql) throws PCJStorageException {
```

Comment:
Should this be makeMetadataDocument(...)? It seems to just be a factory method.

Reply:
well its not really part of a factory, but yeah, i can make it make